src/LibraryManager/CacheService/CacheService.cs

```diff
@@ -129,25 +130,16 @@ private async Task<string> GetResourceAsync(string url, string localFile, int ex
         /// </summary>
         public async Task RefreshCacheAsync(IEnumerable<CacheFileMetadata> librariesCacheMetadata, ILogger logger, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            List<Task> refreshTasks = new List<Task>();
+            await ParallelUtility.ForEachAsync(DownloadFileIfNecessaryAsync, MaxConcurrentDownloads, librariesCacheMetadata, cancellationToken);
 
-            foreach (CacheFileMetadata metadata in librariesCacheMetadata)
+            async Task DownloadFileIfNecessaryAsync(CacheFileMetadata metadata)
             {
-                if (!File.Exists(metadata.DestinationPath) || File.GetLastWriteTime(metadata.DestinationPath) < DateTime.Now.AddDays(-_libraryExpiresAfterDays))
+                if (!File.Exists(metadata.DestinationPath) || File.GetLastWriteTime(metadata.DestinationPath) < DateTime.Now.AddDays(-LibraryExpiresAfterDays))
                 {
-                    var readFileTask = Task.Run(async () =>
```

Comment:
Should Task.Run have moved inside the new ForEachAsync utility code for general case to be correct? See my other comment in the next commit.

Reply:
No.  the new ForEachAsync should not have `Task.Run` as I discussed above.

We used to need it here to add to `refreshTasks` and then do `Task.WhenAll(refreshTasks)` but all that task management was all moved into the `ForEachAsync` helper.